src/cfnlint/rules/resources/HardCodedArnProperties.py

```diff
@@ -0,0 +1,34 @@
+"""
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: MIT-0
+"""
+import re
+from cfnlint.rules import CloudFormationLintRule
+from cfnlint.rules import RuleMatch
+
+
+class HardCodedArnProperties(CloudFormationLintRule):
+    """Checks Resources if ARNs use correctly placed Pseudo Parameters instead of hardcoded Partition, Region, and Account Number"""
+    id = 'I3042'
+    shortdesc = 'ARNs should use correctly placed Pseudo Parameters'
+    description = 'Checks Resources if ARNs use correctly placed Pseudo Parameters instead of hardcoded Partition, Region, and Account Number'
+    source_url = ''
+    tags = ['resources']
+
+    def match(self, cfn):
+        """Check CloudFormation Resources"""
+        matches = []
+        pattern = re.compile(r'arn:(\$\{[^:]*::[^:]*}|[^:]*):[^:]+:(\$\{[^:]*::[^:]*}|[^:]*):(\$\{[^:]*::[^:]*}|[^:]*)')
+        resources = cfn.template.get('Resources', {})
+        if resources:
+            for resourcename, val  in resources.items():
+                candidates = pattern.findall(str(val))
+                for candidate in candidates:
+                    if candidate[0] != '${AWS::Partition}' or candidate[1] not in ('${AWS::Region}', '') or candidate[2] not in ('${AWS::AccountId}', ''):

```

Comment:
We should probably check the length of candidate before we assume it has 3 items.

Reply:
The regex string has 3 capturing groups, so candidate will always have 3 items

Reply:
Those could be. Would it make more sense to have 3 separate rules?

1. Warning: Hardcoded Partition
2. Informational: Hardcoded Region
3. Informational: Hardcoded AccountId

Reply:
I think I could be okay with it being one rule.  I'll admit I've totally messed up the partition portion of this in my templates but for region and account ID I usually do a reasonable job. I have always liked the idea of calling out hard coded values that shouldn't be.  What are your thoughts on if those values were references to parameters?  Or to a FindInMap?

Reply:
That's a good point, I could see the case of trying to specify a cross region resource or a different account as a Parameter. A simple fix would be to take the explicit checks such as `candidate[0] != '${AWS::Region}'`  and check against a regex pattern.

Something similar to the following

```
if candidate[0] != '${AWS::Partition}' or not re.match(r'\$\{\w+}', candidate[1]) or not re.match(r'\$\{\w+}', candidate[2]):
```

Reply:
Just added an update to allow for Parameterized AccountId and Regions

```if candidate[0] != '${AWS::Partition}' or not re.match(r'^\$\{\w+}|\$\{AWS::Region}|$', candidate[1]) or not re.match(r'^\$\{\w+}|\$\{AWS::AccountId}|$', candidate[2]):```